llama-rs/src/lib.rs

```diff
@@ -1296,6 +1316,30 @@ impl Model {
             )
         };
 
+        // Extract logits
+        if let Some(all_logits) = &mut output_request.all_logits {
+            all_logits.resize(n_vocab as usize * n, 0.0);
+            // SAFETY: Data can be read (properly aligned, initialized, data
+            // will not be mutated or otherwise aliased while the slice lives),
+            // and we're not reading past the end of the slice.
+            assert_eq!(input_layer.nelements(), n_vocab * n as i32);
+            let logits_slice = unsafe {
+                slice::from_raw_parts(input_layer.data() as *const f32, n_vocab as usize * n)
```

Comment:
Could you add a function to `Tensor` for getting the data out as a slice? This is what I did:
https://github.com/philpax/ggllama/blob/main/src/ggml.rs#L105-L114

Reply:
Good point!

But, there's a caveat: I don't think this function can be marked as safe. Once something is in a slice, it has to adhere to Rust's aliasing model. The way tensors are implemented, a tensor has to behave like a cheap-to-copy pointer that supports aliasing (this is used in several places of the inference code). This is why tensors have a `share()` function to make this aliasing explicit.

So even if your function tie the lifetime of the tensor to the output slice, there could be many other aliased references to that same tensor through which you could obtain another mutable slice to the same memory. And that's unsound.

What we can do, is have a similar API but mark it as `unsafe` and document the safety requirements for it. But I honestly think it todesn't make sense to try to build an idiomatic safe rust wrapper to ggml. It is a C library built around the idea of aliasing pointers and "being careful" not to drop a context before some other context uses it :sweat_smile:. Even if Rust lifetimes could somehow capture that, the loss in ergonomics would be a high price to pay. One of the main design goals in the original design was to make the code as easy to port as possible. This includes any changes to the tensor code that may happen in the future, like #68.

Any efforts towards making ggml more Rusty would be better directed towards #31.

Reply:
I just realized we already had a similar function for this (`read_data`), so I used that one instead :+1: